compiler/rustc_infer/src/infer/mod.rs

```diff
@@ -1524,25 +1523,40 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
         span: Span,
         lbrct: LateBoundRegionConversionTime,
         value: ty::Binder<'tcx, T>,
-    ) -> (T, BTreeMap<ty::BoundRegion, ty::Region<'tcx>>)
+    ) -> T
     where
-        T: TypeFoldable<'tcx>,
+        T: TypeFoldable<'tcx> + Copy,
     {
-        let fld_r =
-            |br: ty::BoundRegion| self.next_region_var(LateBoundRegion(span, br.kind, lbrct));
-        let fld_t = |_| {
-            self.next_ty_var(TypeVariableOrigin {
-                kind: TypeVariableOriginKind::MiscVariable,
-                span,
+        if let Some(inner) = value.no_bound_vars() {
+            return inner;
+        }
+
+        let mut region_map = FxHashMap::default();
```

Comment:
it looks like the perf regression is caused by us always creating 3 separate storages here and then having to drop them.

Went from `BTreeMap` -> `SsoHashMap` -> `FxHashMap` to figure out if these help. The more complex `Drop` impl (and greater size) of `SsoHashMap` probably made things worse. So hopefully the comparatively simple `HashMap` ends up being an improvement.

If not I think we should commit to tracking bound vars in binders and modify this to eagerly create a `Vec` of inference vars using `value.bound_vars()` and then using that. Doing so should allow us to get some perf improvements in other areas as well.